nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -963,44 +976,40 @@ impl<'a> BlueprintBuilder<'a> {
         allocator.alloc().ok_or(Error::OutOfAddresses { sled_id })
     }
 
-    fn sled_alloc_zpool(
-        &mut self,
+    // Selects a zpools for this zone type.
+    //
+    // This zpool may be used for either durable storage or transient
+    // storage - the usage is up to the caller.
+    //
+    // If `zone_kind` already exists on `sled_id`, it is prevented
+    // from using the same zpool as exisitng zones with the same kind.
+    fn sled_select_zpool(
+        &self,
         sled_id: SledUuid,
-        kind: DiscretionaryOmicronZone,
+        zone_kind: ZoneKind,
     ) -> Result<ZpoolName, Error> {
-        let resources = self.sled_resources(sled_id)?;
+        let all_in_service_zpools =
+            self.sled_resources(sled_id)?.all_zpools(ZpoolFilter::InService);
 
-        // We refuse to choose a zpool for a zone of a given `kind` if this
-        // sled already has a zone of that kind on the same zpool. Build up a
-        // set of invalid zpools for this sled/kind pair.
+        // We refuse to choose a zpool for a zone of a given `zone_kind` if this
+        // sled already has a durable zone of that kind on the same zpool. Build
+        // up a set of invalid zpools for this sled/kind pair.
         let mut skip_zpools = BTreeSet::new();
         for zone_config in self.current_sled_zones(sled_id) {
-            match (kind, &zone_config.zone_type) {
-                (
-                    DiscretionaryOmicronZone::Nexus,
-                    BlueprintZoneType::Nexus(_),
-                ) => {
-                    // TODO handle this case once we track transient datasets
+            if let Some(zpool) = zone_config.zone_type.durable_zpool() {
+                if zone_kind == zone_config.zone_type.kind() {
+                    skip_zpools.insert(zpool);
                 }
-                (
-                    DiscretionaryOmicronZone::CockroachDb,
-                    BlueprintZoneType::CockroachDb(crdb),
-                ) => {
-                    skip_zpools.insert(&crdb.dataset.pool_name);
-                }
-                (DiscretionaryOmicronZone::Nexus, _)
-                | (DiscretionaryOmicronZone::CockroachDb, _) => (),
             }
         }
 
-        for &zpool_id in resources.all_zpools(ZpoolFilter::InService) {
+        for &zpool_id in all_in_service_zpools {
```

Comment:
I know this isn't explicitly part of this PR, but it appears that we always choose the first zpool that doesn't have a zone of the same type on it. This means that we aren't really balancing the load across zpools. I'm not sure if this matters or not, but it seems like it might be somewhat accidental.

Reply:
If I'm reading blame correctly, this disposition was originally made in https://github.com/oxidecomputer/omicron/pull/5797 , when the `sled_alloc_zpool` function was originally added.

@jgallagher - was this to make pool allocation deterministic? I agree that we will probably want the ability to evenly distribute load across different zpools.

Reply:
> @jgallagher - was this to make pool allocation deterministic? 

Yes, and also it was _only_ choosing CRDB zones, so I didn't put much thought into it.

We probably want something like [`OmicronZonePlacement`](https://github.com/oxidecomputer/omicron/blob/45ea9dcab4c8154be360f9ac6e4eee22bdb59387/nexus/reconfigurator/planning/src/planner/omicron_zone_placement.rs#L80-L88) but for disks? (Or to expand `OmicronZonePlacement` to also place disks.) It is deterministic but also distributes load as evenly as it can within the set of restrictions it has. I'll file an issue.

Reply:
Makes total sense, thanks for filing. I can look into the issue.